drivers/network/netio/netio.c

```diff
@@ -0,0 +1,860 @@
+#include <tdikrnl.h>
+#include <tdiinfo.h>
+#include <tdi_proto.h>
+#include <tdiconn.h>
+
+/* AFD also defines this. It is used by the tdi helpers */
+DWORD DebugTraceLevel = MIN_TRACE;
+
+#ifndef TAG_AFD_TDI_CONNECTION_INFORMATION
+#define TAG_AFD_TDI_CONNECTION_INFORMATION 'cTfA'
+#endif
+
+/* AFD Share Flags */
+#define AFD_SHARE_UNIQUE    0x0L
+#define AFD_SHARE_REUSE     0x1L
+#define AFD_SHARE_WILDCARD  0x2L
+#define AFD_SHARE_EXCLUSIVE 0x3L
+
+struct _WSK_SOCKET_INTERNAL
+{
+    struct _WSK_SOCKET s;
+
+    ADDRESS_FAMILY family;      /* AF_INET or AF_INET6 */
+    unsigned short type;        /* SOCK_DGRAM, SOCK_STREAM, ... */
+    unsigned long proto;        /* IPPROTO_UDP, IPPROTO_TCP */
+    unsigned long flags;        /* WSK_FLAG_LISTEN_SOCKET, ... */
+    void *user_context;         /* parameter for callbacks, opaque */
+    struct _UNICODE_STRING TdiName;     /* \\Devices\\Tcp, \\Devices\\Udp */
+
+    struct sockaddr LocalAddress;
+    PFILE_OBJECT LocalAddressFile;
+    HANDLE LocalAddressHandle;
+
+    struct sockaddr RemoteAddress;      /* Remeber latest remote connection */
+
+    /* Those exist for connection oriented (TCP/IP) sockets only */
+    PFILE_OBJECT ConnectionFile;        /* Returned by TdiOpenConnectionEndpointFile() */
+    HANDLE ConnectionHandle;    /* Returned by TdiOpenConnectionEndpointFile() */
+
+    /* Incoming connection callback function: */
+    const struct _WSK_CLIENT_LISTEN_DISPATCH *ListenDispatch;
+    int CallbackMask;
+
+    int Flags;                  /* SO_REUSEADDR, ... see ws2def.h */
+    int RefCount;               /* See SocketGet/SocketPut TODO: this should be atomic */
+};
+
+struct NetioContext
+{
+    PIRP UserIrp;
+    struct _WSK_SOCKET_INTERNAL *socket;
+};
+
+void
+SocketGet(struct _WSK_SOCKET_INTERNAL *s)
+{
+    s->RefCount++;
+}
+
+void
+SocketPut(struct _WSK_SOCKET_INTERNAL *s)
+{
+    s->RefCount--;
+    if (s->RefCount == 0)
+    {
+        if (s->LocalAddressHandle != NULL)
+        {
+            ZwClose(s->LocalAddressHandle);
+            s->LocalAddressHandle = NULL;
+            s->LocalAddressFile = NULL;
+        }
+        if (s->ConnectionHandle != NULL)
+        {
+            ZwClose(s->ConnectionHandle);
+            s->ConnectionHandle = NULL;
+            s->ConnectionFile = NULL;
+        }
+        ExFreePool(s);
+    }
+}
+
+NTSTATUS NTAPI
+DriverEntry(PDRIVER_OBJECT DriverObject, PUNICODE_STRING RegistryPath)
+{
+    DbgPrint("netio.sys DriverEntry compiled " __DATE__ " " __TIME__ "\n");
+
+    return STATUS_SUCCESS;
+}
+
+static NTSTATUS NTAPI
+NetioComplete(PDEVICE_OBJECT DeviceObject, PIRP Irp, PVOID Context)
+{
+    struct NetioContext *c = (struct NetioContext *)Context;
+    PIRP UserIrp = c->UserIrp;
+
+    UserIrp->IoStatus.Status = Irp->IoStatus.Status;
+    UserIrp->IoStatus.Information = Irp->IoStatus.Information;
+
+    IoCompleteRequest(UserIrp, IO_NETWORK_INCREMENT);
+
+    SocketPut(c->socket);
+    ExFreePoolWithTag(c, 'NEIO');
```

Comment:
It may be of interest to have a define for this tag.
Also it should certainly be `'OIEN'`.

Reply:
Ok done.